pkg/utils/common/common.go

```diff
@@ -135,53 +135,62 @@ func ResolveWatchNamespaceForManager(oprNamespace string, watchNamespace string)
 	return false, strings.Split(watchNamespace, ",")
 }
 
-type BrokerVersion struct {
-	Version *semver.Version
-}
+func ResolveBrokerVersion(versions []semver.Version, desired string) *semver.Version {
 
-func NewBrokerVersion(ver *semver.Version) *BrokerVersion {
-	return &BrokerVersion{
-		Version: ver,
+	if len(versions) == 0 {
+		return nil
+	}
+	if desired == "" {
+		// latest
+		return &versions[len(versions)-1]
 	}
-}
-
-func (bver *BrokerVersion) MajorOnly() bool {
-	fields := strings.Split(bver.Version.Original(), ".")
-	return len(fields) == 1
-}
-
-func (bver *BrokerVersion) MinorOnly() bool {
-	fields := strings.Split(bver.Version.Original(), ".")
-	return len(fields) == 2
-}
 
-func (bver *BrokerVersion) IsMatch(ver *semver.Version) bool {
-	if bver.MajorOnly() {
-		return bver.Version.Major() == ver.Major()
+	major, minor, patch := resolveVersionComponents(desired)
+
+	// walk the ordered tree in reverse, locking down match based on desired version components
+	var i int = len(versions) - 1
+	for ; i >= 0; i-- {
+		if major != nil {
+			if *major == versions[i].Major {
+				if minor == nil {
+					break
+				} else if *minor == versions[i].Minor {
+					if patch == nil {
+						break
+					} else if *patch == versions[i].Patch {
+						break
+					}
+				}
+			}
+		}
 	}
```

Comment:
The semver range could simplify the search
```suggestion
	// Build matchDesiredRange from desired
	matchDesiredRange := semver.MustParseRange(">=0.0.0")
	if desired != "" {
		var parseErr error

		// Add the wildcard suffix if the semantic version is incomplete
		if strings.Count(desired, ".") < 2 && !strings.HasSuffix(desired, ".x") {
			matchDesiredRange, parseErr = semver.ParseRange(desired + ".x")
		} else {
			matchDesiredRange, parseErr = semver.ParseRange(desired)
		}

		if parseErr != nil {
			return nil
		}
	}

	for i := len(versions) - 1; i >= 0; i-- {
		if matchDesiredRange(versions[i]) {
			return &versions[i]
		}
	}
```

Reply:
am not convinced, I see it as a match/pruning process rather than a range issue. Maybe if we supported ranges in Version it would make sense. 
But thanks for the suggestion. We may use it in the future.

Reply:
or are you thinking that our version list may contain patch and descriptive semantic versions in the future. like 2.20.7-pre-01 or some other semver format and we are future proofing?

Reply:
I suggested this change only because it seems we are doing the same thing of blang semver range but I see your point, I don't have a strong opinion on that so feel free to reject this suggestion.

Reply:
thanks for the feedback, much appreciated.